### ISSUE TITLE
Cycle detected when evaluating trait selection obligation

### ISSUE THREAD

**Issue by zhoukezi** [google/zerocopy#2116]
With `zerocopy` 0.8.11, the following code failed to compile:

```rust
use zerocopy::*;

#[derive(KnownLayout)]
#[repr(C)]
pub struct Dummy([u8; Self::CONST]);

impl Dummy {
    const CONST: usize = 42;
}
```

```plain
error[E0391]: cycle detected when evaluating trait selection obligation `_::__ZerocopyKnownLayoutMaybeUninit: core::iter::traits::collect::IntoIterator`
  |
note: ...which requires evaluating type-level constant...
 --> src/lib.rs:5:23
  |
5 | pub struct Dummy([u8; Self::CONST]);
  |                       ^^^^^^^^^^^
note: ...which requires const-evaluating + checking `_::__ZerocopyKnownLayoutMaybeUninit::0::{constant#0}`...
 --> src/lib.rs:5:23
  |
5 | pub struct Dummy([u8; Self::CONST]);
  |                       ^^^^^^^^^^^
note: ...which requires caching mir of `_::__ZerocopyKnownLayoutMaybeUninit::0::{constant#0}` for CTFE...
 --> src/lib.rs:5:23
  |
5 | pub struct Dummy([u8; Self::CONST]);
  |                       ^^^^^^^^^^^
note: ...which requires elaborating drops for `_::__ZerocopyKnownLayoutMaybeUninit::0::{constant#0}`...
 --> src/lib.rs:5:23
  |
5 | pub struct Dummy([u8; Self::CONST]);
  |                       ^^^^^^^^^^^
note: ...which requires borrow-checking `_::__ZerocopyKnownLayoutMaybeUninit::0::{constant#0}`...
 --> src/lib.rs:5:23
  |
5 | pub struct Dummy([u8; Self::CONST]);
  |                       ^^^^^^^^^^^
note: ...which requires promoting constants in MIR for `_::__ZerocopyKnownLayoutMaybeUninit::0::{constant#0}`...
 --> src/lib.rs:5:23
  |
5 | pub struct Dummy([u8; Self::CONST]);
  |                       ^^^^^^^^^^^
note: ...which requires const checking `_::__ZerocopyKnownLayoutMaybeUninit::0::{constant#0}`...
 --> src/lib.rs:5:23
  |
5 | pub struct Dummy([u8; Self::CONST]);
  |                       ^^^^^^^^^^^
note: ...which requires building MIR for `_::__ZerocopyKnownLayoutMaybeUninit::0::{constant#0}`...
 --> src/lib.rs:5:23
  |
5 | pub struct Dummy([u8; Self::CONST]);
  |                       ^^^^^^^^^^^
note: ...which requires match-checking `_::__ZerocopyKnownLayoutMaybeUninit::0::{constant#0}`...
 --> src/lib.rs:5:23
  |
5 | pub struct Dummy([u8; Self::CONST]);
  |                       ^^^^^^^^^^^
note: ...which requires type-checking `_::__ZerocopyKnownLayoutMaybeUninit::0::{constant#0}`...
 --> src/lib.rs:5:29
  |
5 | pub struct Dummy([u8; Self::CONST]);
  |                             ^^^^^
  = note: ...which again requires evaluating trait selection obligation `_::__ZerocopyKnownLayoutMaybeUninit: core::iter::traits::collect::IntoIterator`, completing the cycle
note: cycle used when type-checking `_::<impl at src/lib.rs:3:10: 3:21>::{constant#0}`
 --> src/lib.rs:5:29
  |
5 | pub struct Dummy([u8; Self::CONST]);
  |                             ^^^^^
  = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information

For more information about this error, try `rustc --explain E0391`.
```

This compiles on 0.8.10.

```
> rustc --version
rustc 1.83.0 (90b35a623 2024-11-26)
> cargo --version
cargo 1.83.0 (5ffbef321 2024-10-29)
> cat Cargo.toml
[package]
name = "sandbox"
version = "0.1.0"
edition = "2021"

[dependencies]
#zerocopy = { version = "=0.8.11", features = ["derive"] }
zerocopy = { version = "=0.8.10", features = ["derive"] }
```

**Comment by joshlf** [comment]
Thanks for reporting this! We'll take a look.

**Comment by jswrenn** [comment]
This is definitely a bug on our end, though the rustc error message is also a red herring. Here's a self-contained repro:
```rust
trait KnownLayout {
    type MaybeUninit;
}

impl KnownLayout for u8 {
    type MaybeUninit = core::mem::MaybeUninit<Self>;
}

impl<const N: usize, T> KnownLayout for [T; N]
where
    T: KnownLayout
{
    type MaybeUninit = [<T as KnownLayout>::MaybeUninit; N];
}

#[repr(C)]
pub struct Dummy([u8; Self::CONST]);

impl Dummy {
    const CONST: usize = 42;
}

const _: () = {
    #[repr(C)]
    struct DummyMaybeUninit(<[u8; Self::CONST] as KnownLayout>::MaybeUninit);

    impl KnownLayout for Dummy {
        type MaybeUninit = DummyMaybeUninit;
    }
};
```

The issue, here, is this line:

```rust
struct DummyMaybeUninit(<[u8; Self::CONST] as KnownLayout>::MaybeUninit);
```

...because `Self` takes on a different type when emplaced in `DummyMaybeUninit`. The `DummyMaybeUninit` doesn't have an associated `CONST`, and so this code contains a type error.

**Comment by jswrenn** [comment]
We can probably fix this by replacing occurrences of `Self` with `#ident #ty_generics` in `__ZerocopyKnownLayoutMaybeUninit`.

**Comment by jswrenn** [comment]
Hm, that approach is complicated by the fact that const exprs can be arbitrarily complicated and define new types in which `Self` takes on a different meaning; e.g.:

```rust


struct Headache([u8; {
    struct Migraine;

    impl Migraine {
        const A: usize = 21;
        const B: usize = 21;
        const LEN: usize = Self::A + Self::B;
    }

    impl Headache {
        const A: usize = Migraine::LEN;
        const B: usize = Migraine::LEN;
        const LEN: usize = Self::A + Self::B;
    }

    Self::LEN
}]);
```

**Comment by jswrenn** [comment]
Perhaps, instead, we can add a `TrailingField` associated type to `KnownLayout` and reference that instead. It's a layer of indirection that will make the safety proofs more annoying, but it should do the trick.

**Comment by joshlf** [comment]
Fixed by #2127 and published in [0.8.13](https://crates.io/crates/zerocopy/0.8.13).
